src/generic/numberparsing.h

```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
+    0xa7709a56, 0x88a66076, 0x6acff893, 0x583f6b8,  0xc3727a33, 0x744f18c0,
+    0x1162def0, 0x8addcb56, 0x6d953e2b, 0xc8fa8db6, 0x1d9c9892, 0x2503beb6,
+    0x2e44ae64, 0x5ceaecfe, 0x7425a83e, 0xd12f124e, 0x82bd6b70, 0x636cc64d,
+    0x3c47f7e0, 0x65acfaec, 0x7f1839a7, 0x1ede4811, 0x934aed0a, 0xf81da84d,
+    0x36251260, 0xc1d72b7c, 0xb24cf65b, 0xdee033f2, 0x169840ef, 0x8e1f2895,
+    0xf1a6f2ba, 0xae10af69, 0xacca6da1, 0x17fd090a, 0xddfc4b4c, 0x4abdaf10,
+    0x9d6d1ad4, 0x84c86189, 0x32fd3cf5, 0x3fbc8c33, 0xfabaf3f,  0x29cb4d87,
+    0x743e20e9, 0x914da924, 0x1ad089b6, 0xa184ac24, 0xc9e5d72d, 0x7e2fa67c,
+    0xddbb901b, 0x552a7422, 0xd53a8895, 0x8a892aba, 0x2d2b7569, 0x9c3b2962,
+    0x8349f3ba, 0x241c70a9, 0xed238cd3, 0xf4363804, 0xb143c605, 0xdd94b786,
+    0xca7cf2b4, 0xfd1c2f61, 0xbc633b39, 0xd5be0503, 0x4b2d8644, 0xddf8e7d6,
+    0xcabb90e5, 0x3d6a751f, 0xcc51267,  0x27fb2b80, 0xb1f9f660, 0x5e7873f8,
+    0xdb0b487b, 0x91ce1a9a, 0x7641a140, 0xa9e904c8, 0x546345fa, 0xa97c1779,
+    0x49ed8eab, 0x5c68f256, 0x73832eec, 0xc831fd53, 0xba3e7ca8, 0x28ce1bd2,
+    0x7980d163, 0xd7e105bc, 0x8dd9472b, 0xb14f98f6, 0x6ed1bf9a, 0xa862f80,
+    0xcd27bb61, 0x8038d51c, 0xe0470a63, 0x1858ccfc, 0xf37801e,  0xd3056025,
+    0x47c6b82e, 0x4cdc331d, 0xe0133fe4, 0x58180fdd, 0x570f09ea};
 
-// called by parse_number when we know that the output is a float,
-// but where there might be some integer overflow. The trick here is to
-// parse using floats from the start.
-// Do not call this function directly as it skips some of the checks from
-// parse_number
-//
-// This function will almost never be called!!!
-//
-// Note: a redesign could avoid this function entirely.
-//
-static never_inline bool parse_float(const uint8_t *const buf, ParsedJson &pj,
-                                     const uint32_t offset, bool found_minus) {
-  const char *p = reinterpret_cast<const char *>(buf + offset);
-  bool negative = false;
-  if (found_minus) {
-    ++p;
-    negative = true;
+// handle case where strtod finds an invalid number. won't we have a buffer
+// overflow if it's just numbers past the end?
+// the assumption is that power_index is in the [0, 2*308] interval
+static really_inline double compute_float_64(uint64_t power_index, uint64_t i,
+                                             bool negative, bool *success) {
+  components c =
+      power_of_ten_components[power_index]; // safe because power_index <= 2*308
+  uint64_t factor_mantissa = c.mantissa;
+  int lz = leading_zeroes(i);
+  i <<= lz;
+  __uint128_t large_mantissa = (__uint128_t)i * factor_mantissa;
+  uint64_t upper = large_mantissa >> 64;
+  if (unlikely((upper & 0x1FF) == 0x1FF)) {
+    *success = false;
+    return 0;
   }
-  long double i;
-  if (*p == '0') { // 0 cannot be followed by an integer
-    ++p;
-    i = 0;
+  uint64_t mantissa = 0;
+  if (upper & (1ULL << 63)) { // todo: this branch needs to be explained
```

Comment:
we've passed the `return 0`, so we know that what we have will not be affected by the inaccuracy of `f`, and now we just need to do bookkeeping. is the MSB 1? if so, shift by 10 instead of 9